### PR TITLE
Adding EGT Model for Graph Classification

### PR DESCRIPTION
**Pull request by rudongyu**
# What does this PR do?

Add the graph transformer model EGT for the graph classification task.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Hey! Thanks a lot for wanting to contribute! I would recommend you to push your model on the hub following[ this tutorial](https://huggingface.co/docs/transformers/custom_models) as it will be a lot easier and doesn't require you to go through all the CI and should be easier to do! Feel free to then share the huggingface repository were you added it!

**Comment by ZHITENGLI** [comment]
> Hey! Thanks a lot for wanting to contribute! I would recommend you to push your model on the hub following[ this tutorial](https://huggingface.co/docs/transformers/custom_models) as it will be a lot easier and doesn't require you to go through all the CI and should be easier to do! Feel free to then share the huggingface repository were you added it!

Thanks for your suggestions! I have pushed our model on the hub and [this](https://huggingface.co/Zhiteng/dgl-egt/tree/main) is the huggingface repository where I added it.

**Comment by rudongyu** [comment]
> Hey! Thanks a lot for wanting to contribute! I would recommend you to push your model on the hub following[ this tutorial](https://huggingface.co/docs/transformers/custom_models) as it will be a lot easier and doesn't require you to go through all the CI and should be easier to do! Feel free to then share the huggingface repository were you added it!

Thanks for your kind reminder! I went through the tutorial and found two potential barriers to directly sharing models in that manner:
1. HF transformers doesn't have a definition for AutoModelForGraphClassification.
2. The tutorial doesn't include a way for sharing the data collating method. However, it could be a critical part for users conveniently applying the model for graph classification.

So, maybe providing the model in the library is still a better way in this case? @clefourrier for awareness.

File: src/transformers/models/egt/collating_egt.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+from typing import Any, Dict, List, Mapping
+
+import numpy as np
+import torch
+
+from ...utils import is_dgl_available, requires_backends
+
+
+if is_dgl_available():
+    import dgl
```
Comment (clefourrier) [inline src/transformers/models/egt/collating_egt.py:10]: Hi, we try to reduce the number of dependencies needed for transformers, please import the related functions in a specific file

File: src/transformers/models/egt/collating_egt.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+def convert_to_single_edge_emb(x, offset: int = 8):
+    feature_num = x.shape[1] if len(x.shape) > 1 else 1
+    feature_offset = 1 + np.arange(0, feature_num * offset, offset, dtype=np.int64)
+    x = x + feature_offset
+    return x
+
+
+def preprocess_item(item, keep_features=True):
+    requires_backends(preprocess_item, ["dgl"])
```
Comment (clefourrier) [inline src/transformers/models/egt/collating_egt.py:28]: same here, remove this

File: src/transformers/models/egt/modeling_egt.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    "Zhiteng/dgl-egt",
+]
+
+
+class VirtualNodes(nn.Module):
+    """
+    Generate node and edge features for virtual nodes in the graph and pad the corresponding matrices.
+    """
+
+    def __init__(self, feat_size, edge_feat_size, num_virtual_nodes=1):
```
Comment (clefourrier) [inline src/transformers/models/egt/modeling_egt.py:41]: You need docstrings everywhere to explain what the arguments are, their types, and what each function does

File: src/transformers/models/egt/modeling_egt.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        self.feat_size = feat_size
+        self.edge_feat_size = edge_feat_size
+        self.num_virtual_nodes = num_virtual_nodes
+
+        self.vn_node_embeddings = nn.Parameter(torch.empty(num_virtual_nodes, self.feat_size))
+        self.vn_edge_embeddings = nn.Parameter(torch.empty(num_virtual_nodes, self.edge_feat_size))
+        nn.init.normal_(self.vn_node_embeddings)
+        nn.init.normal_(self.vn_edge_embeddings)
+
+    def forward(self, h, e, mask):
```
Comment (clefourrier) [inline src/transformers/models/egt/modeling_egt.py:52]: h, e, etc are not explicit argument names, please change them

File: src/transformers/models/egt/modeling_egt.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    model with a downstream model of your choice, following the example in EGTForGraphClassification.
+    """
+
+    def __init__(self, config: EGTConfig):
+        super().__init__(config)
+
+        self.activation = getattr(nn, config.activation)()
+
+        self.layer_common_kwargs = {
+            "feat_size": config.feat_size,
```
Comment (clefourrier) [inline src/transformers/models/egt/modeling_egt.py:108]: are those features node features or graph features?

File: src/transformers/models/egt/collating_egt.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        node_feat_size = len(features[0]["input_nodes"][0])
+        edge_feat_size = len(features[0]["attn_edge_type"][0][0])
+        svd_pe_size = len(features[0]["svd_pe"][0]) // 2
+        batch_size = len(features)
+
+        batch["featm"] = torch.zeros(batch_size, max_node_num, max_node_num, edge_feat_size, dtype=torch.long)
+        batch["dm"] = torch.zeros(batch_size, max_node_num, max_node_num, dtype=torch.long)
+        batch["node_feat"] = torch.zeros(batch_size, max_node_num, node_feat_size, dtype=torch.long)
+        batch["svd_pe"] = torch.zeros(batch_size, max_node_num, svd_pe_size * 2, dtype=torch.float)
+        batch["attn_mask"] = torch.zeros(batch_size, max_node_num, dtype=torch.long)
```
Comment (clefourrier) [inline src/transformers/models/egt/collating_egt.py:88]: again, these keys are not explicit enough names

File: src/transformers/models/egt/configuration_egt.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    # See all Graphormer models at https://huggingface.co/models?filter=graphormer
+}
+
+
+class EGTConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~EGTModel`]. It is used to instantiate an EGT
+    model according to the specified arguments, defining the model architecture. Instantiating a configuration with the
+    defaults will yield a similar configuration to that of the EGT
+    [graphormer-base-pcqm4mv1](https://huggingface.co/graphormer-base-pcqm4mv1) architecture.
```
Comment (clefourrier) [inline src/transformers/models/egt/configuration_egt.py:21]: Why is a Graphormer your default pretrained model? I'm quite sure the architectures are not even compatible

**Review by clefourrier — COMMENTED** [review]
Hi, just a took a first look - this PR looks nice but 
- you need to add considerably more docstrings and type hinting 
- change variable names to make them more explicit (please no "one letter names" unless it's i as index in a for loop)
- we usually don't add more dependencies to transformers (@ArthurZucker for confirmation?)

Are you providing a base pretrained model hosted on the hub with? I'm very confused as to why the base model you provide is a link to a Graphormer one

**Pull request closed without merging**